File: General-DSA-Questions/Linked-List/15_2ImpQues.cpp
// Question 1 : Find intersection of Two Linked Lists
// Problem Statement: Given the heads of two singly linked-lists headA and headB, return the node at which the two lists intersect.
// If the two linked lists have no intersection at all, return null.

// Intiution:
// 1. Calculate the lengths of both linked lists. Say the lengths are lenA and lenB.
// 2. Find the difference in lengths, say d = |lenA - lenB|.
// 3. Move the pointer of the longer list d nodes ahead.
// 4. Now, move both pointers one node at a time until they meet. The meeting point is the intersection node.
// 5. If they reach the end (nullptr) without meeting, then there is no intersection.



// Question 2 : Rotate a Linked List
// Problem Statement: Given the head of a linked list, rotate the list to the right by k places and return the new head of the rotated list.

// Intiution:
// 1. Calculate the length of the linked list, say len.
// 2. If k >= len, then do k = k % len to handle cases where k is greater than the length of the list.
// 3. If k is 0 after the above step, return the head as no rotation is needed.
// 4. Find the (len - k)th node in the list. This will be the new tail after rotation.
// 5. The next node of the new tail will be the new head.
// 6. Set the next of the current tail to the current head to make it circular.
// 7. Set the next of the new tail to nullptr to break the circularity.
// 8. Return the new head.

File: JS-DSA-Questions/Linked-List/15_2ImpQues.js
// Two important linked-list problems (theory + reference implementations).

// =====================================================================================
// Question 1 : Find the intersection of two linked lists
// =====================================================================================
// Given the heads of two singly linked lists headA and headB, return the node at which they
// intersect, or null if they never intersect.
//
// Intuition:
// 1. Compute lengths lenA and lenB.
// 2. d = |lenA - lenB|. Advance the longer list's pointer d nodes.
// 3. Move both pointers together until they meet — that node is the intersection.
// 4. If they reach the end (null) without meeting, there is no intersection.

class Node {
  constructor(val) {
    this.data = val;
    this.next = null;
  }
}

function getLength(head) {
  let len = 0;
  while (head !== null) {
    len++;
    head = head.next;
  }
  return len;
}

function getIntersectionNode(headA, headB) {
  let lenA = getLength(headA);
  let lenB = getLength(headB);

  let a = headA;
  let b = headB;

  // Advance the longer list's pointer by the length difference
  while (lenA > lenB) {
    a = a.next;
    lenA--;
  }
  while (lenB > lenA) {
    b = b.next;
    lenB--;
  }

  // Move together until they meet (or both reach null)
  while (a !== b) {
    a = a.next;
    b = b.next;
  }
  return a; // intersection node or null
}

// =====================================================================================
// Question 2 : Rotate a linked list to the right by k places
// =====================================================================================
// Intuition:
// 1. Compute length len.
// 2. k = k % len (handles k >= len). If k === 0, no rotation needed.
// 3. The new tail is the (len - k)th node; the node after it becomes the new head.
// 4. Make the list circular (old tail -> old head), then break it at the new tail.

function rotateRight(head, k) {
  if (head === null || head.next === null || k === 0) return head;

  // Compute length and find the current tail
  let len = 1;
  let tail = head;
  while (tail.next !== null) {
    tail = tail.next;
    len++;
  }

  k = k % len;
  if (k === 0) return head;

  // Make the list circular
  tail.next = head;

  // New tail is at position (len - k); new head is the node after it
  let stepsToNewTail = len - k;
  let newTail = head;
  for (let i = 1; i < stepsToNewTail; i++) {
    newTail = newTail.next;
  }
  const newHead = newTail.next;
  newTail.next = null; // break the circle

  return newHead;
}

// Demo for rotation: 1 -> 2 -> 3 -> 4 -> 5, k = 2  →  4 -> 5 -> 1 -> 2 -> 3
let head = new Node(1);
head.next = new Node(2);
head.next.next = new Node(3);
head.next.next.next = new Node(4);
head.next.next.next.next = new Node(5);

let rotated = rotateRight(head, 2);
const out = [];
while (rotated) {
  out.push(rotated.data);
  rotated = rotated.next;
}
console.log(out.join(" ")); // 4 5 1 2 3
